iohub/ngff_utils.py

```diff
@@ -0,0 +1,341 @@
+    Multiprocessing over T only:
+    - input_channel_idx and output_channel_idx should be provided.
+    """
+    # Function to be applied
+    click.echo(f"Function to be applied: \t{func}")
+
+    # Get the reader and writer
+    click.echo(f"Input data path:\t{input_data_path}")
+    click.echo(f"Output data path:\t{str(output_path)}")
+    input_dataset = open_ome_zarr(str(input_data_path))
+    stdout_buffer = io.StringIO()
+    with contextlib.redirect_stdout(stdout_buffer):
+        input_dataset.print_tree()
+    click.echo(f" Input data tree: {stdout_buffer.getvalue()}")
+
+    # Find time indices
+    if time_indices_in == "all":
+        time_indices_in = range(input_dataset.data.shape[0])
+        time_indices_out = time_indices_in
+    elif isinstance(time_indices_in, list):
+        # Check for invalid times
+        time_ubound = input_dataset.data.shape[0] - 1
+        if np.max(time_indices_in) > time_ubound:
+            raise ValueError(
+                f"time_indices_in = {time_indices_in} includes a time index beyond the maximum index of the dataset = {time_ubound}"
+            )
+        # Handle the case when time_indices out is not provided. It defaults to the t_indices_in
+        if len(time_indices_out) == 0:
+            time_indices_out = range(len(time_indices_in))
+
+    # Check the arguments for the function
+    all_func_params = inspect.signature(func).parameters.keys()
+    # Extract the relevant kwargs for the function 'func'
+    func_args = {}
+    non_func_args = {}
+
+    for k, v in kwargs.items():
+        if k in all_func_params:
+            func_args[k] = v
+        else:
+            non_func_args[k] = v
+
+    # Write the settings into the metadata if existing
+    if "extra_metadata" in non_func_args:
+        # For each dictionary in the nest
+        with open_ome_zarr(output_path, mode="r+") as output_dataset:
+            for params_metadata_keys in kwargs["extra_metadata"].keys():
+                output_dataset.zattrs["extra_metadata"] = non_func_args[
+                    "extra_metadata"
+                ]
+
+    # Loop through (T, C), deskewing and writing as we go
+    click.echo(f"\nStarting multiprocess pool with {num_processes} processes")
+
+    if input_channel_idx is None or len(input_channel_idx) == 0:
+        # If C is not empty, use itertools.product with both ranges
+        _, C, _, _, _ = input_dataset.data.shape
+        iterable = [
+            ([c], [c], time_idx, time_idx_out)
+            for (time_idx, time_idx_out), c in itertools.product(
+                zip(time_indices_in, time_indices_out), range(C)
+            )
+        ]
+        partial_apply_transform_to_zyx_and_save = partial(
+            apply_transform_to_zyx_and_save,
+            func,
+            input_dataset,
+            output_path / Path(*input_data_path.parts[-3:]),
+            **func_args,
+        )
+    else:
+        # If C is empty, use only the range for time_indices_in
```

Comment:
With good structure for input and output T, C, and Z indices, we should be able to simplify this to:

```python
for (t_idx_in, c_idx_in, z_idx_in), (t_idx_out, c_idx_out, z_idx_out) in zip(
    itertools.product(input_time_indices, input_channel_indices, input_slice_indices),
    itertools.product(output_time_indices, output_channel_indices, output_slice_indices),
)
    output[t_idx_out, c_idx_out, z_idx_out] = func(input[t_idx_in, c_idx_in, z_idx_in])
```

Reply:
Addressed in person. We're doing `itertools` product over t and c, but not z.